core/src/main/scala/chisel3/internal/chisel-circt-binder/src/PanamaCIRCT.scala

```diff
@@ -0,0 +1,394 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.internal.panama.circt
+
+import java.lang.foreign._
+import java.lang.foreign.MemorySegment.NULL
+import java.lang.foreign.ValueLayout._
+
+import org.llvm.circt
+import org.llvm.circt.CAPI
+
+// Wrapper for CIRCT APIs with Panama framework
+class PanamaCIRCT {
+  // Open an arena for memory management of MLIR API calling in this context instance
+  private val arena = Arena.openConfined()
+
+  // Create MLIR context and register dialects we need
+  private val mlirCtx = {
+    val mlirCtx = CAPI.mlirContextCreate(arena)
+
+    // Register dialects
+    Seq(
+      CAPI.mlirGetDialectHandle__firrtl__(arena),
+      CAPI.mlirGetDialectHandle__chirrtl__(arena)
+    ).foreach(CAPI.mlirDialectHandleLoadDialect(arena, _, mlirCtx))
+
+    mlirCtx
+  }
+
+  // Public items for outside
+
+  // Constants for this instance
+  val unkLoc = mlirLocationUnknownGet()
+  val emptyArrayAttr = mlirArrayAttrGet(Seq.empty)
+
+  //////////////////////////////////////////////////
+  // Helpers
+  //
+
+  private def newString(string: String): MlirStringRef = {
+    val bytes = string.getBytes()
+    val buffer = MemorySegment.allocateNative(bytes.length + 1, arena.scope())
+    buffer.copyFrom(MemorySegment.ofArray(bytes))
+    MlirStringRef(CAPI.mlirStringRefCreateFromCString(arena, buffer))
+  }
+
+  private def seqToArray[T <: ForeignType[_]](xs: Seq[T]): (MemorySegment, Int) = {
+    if (xs.nonEmpty) {
+      val sizeOfT = xs(0).sizeof
+
+      val buffer = MemorySegment.allocateNative(sizeOfT * xs.length, arena.scope())
+      xs.zipWithIndex.foreach {
+        case (x, i) =>
+          x.get match {
+            case value: MemorySegment => buffer.asSlice(sizeOfT * i, sizeOfT).copyFrom(value)
+            case value: Int           => buffer.setAtIndex(CAPI.C_INT, i, value)
+          }
+      }
+      (buffer, xs.length)
+    } else {
+      (NULL, 0)
+    }
+  }
+
+  //////////////////////////////////////////////////
+  // CIRCT APIs
+  //
+
+  def mlirModuleCreateEmpty(location: MlirLocation) = MlirModule(CAPI.mlirModuleCreateEmpty(arena, location.get))
+
+  def mlirModuleGetBody(module: MlirModule) = MlirBlock(CAPI.mlirModuleGetBody(arena, module.get))
+
+  def mlirModuleGetOperation(module: MlirModule) = MlirOperation(CAPI.mlirModuleGetOperation(arena, module.get))
+
+  def mlirOperationStateGet(name: String, loc: MlirLocation) = MlirOperationState(
+    CAPI.mlirOperationStateGet(arena, newString(name).get, loc.get)
+  )
+
+  def mlirOperationStateAddAttributes(state: MlirOperationState, attrs: Seq[MlirNamedAttribute]) = {
+    if (attrs.nonEmpty) {
```

Comment:
learn to use `attrs.foreach`

Reply:
Here we convert `attrs` to C array `(MemorySegment, Int)`, `.foreach` is not helpful.